### PR TITLE
HPCC-19784 Improve access to user and resource name value pairs

### PR DESCRIPTION
**Pull request by timothyklemm**
Expose iterators to ISecUser's properties and ISecResource's
parameters. Without these iterators, the data can only be
accessed within the concrete implementation class or by code
that knows exactly what names to look for. With the iterators,
the data may be manipulated by generic code that only knows
that some arbitrary data may be present.

Fix HPCC-19784

Signed-off-by: Tim Klemm <Tim.Klemm@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by timothyklemm** [comment]
@RussWhitehead @mayx Please review.

File: system/security/LdapSecurity/ldapsecurity.ipp [hunk 1]
```diff
@@ -149,6 +149,7 @@ public:
 
     virtual void setPropertyInt(const char* name, int value){}
     virtual int getPropertyInt(const char* name){ return 0;}
+    IPropertyIterator * getPropertyIterator() const override { return nullptr;}
```
Comment (RussWhitehead) [inline system/security/LdapSecurity/ldapsecurity.ipp]: Should be virtual ?

**Review by RussWhitehead — CHANGES_REQUESTED** [review]
Just one minor comment

File: system/security/LdapSecurity/ldapsecurity.ipp (hunk 1 above)
Comment (timothyklemm) [reply system/security/LdapSecurity/ldapsecurity.ipp]: I will leave this as-is. Per Tony Fishbeck:

The reason it is ok to use the override keyword even when the surrounding code doesn’t is that that is an improvement we are trying to make slowly throughout our codebase whenever we make changes to existing code.  It is a very good practice and therefore we think we should use it even when changing older code.

I don’t think you need to add the virtual keyword.  The virtual keyword used to serve as a sort of documentation even if the function was already virtual.  Now the override keyword serves that purpose as well.

**Comment by timothyklemm** [comment]
@rpastrana Can you sign off on this since Russ is not available at the moment?

File: system/security/LdapSecurity/ldapsecurity.ipp (hunk 1 above)
Comment (rpastrana) [reply system/security/LdapSecurity/ldapsecurity.ipp]: @timothyklemm the lone override is fine, for consistency's sake, if  the function is to be virtual, go ahead and label it as such, even if it's redundant.

Comment (rpastrana) [reply system/security/LdapSecurity/ldapsecurity.ipp]: @timothyklemm your verbal explanation makes sense.

**Review by rpastrana — APPROVED** [review]
@timothyklemm looks good

**Comment by timothyklemm** [comment]
@richardkchapman Please merge.

**Pull request merged**